confignode/src/main/java/org/apache/iotdb/confignode/procedure/impl/RemoveConfigNodeProcedure.java

```diff
@@ -0,0 +1,177 @@
+import org.apache.iotdb.confignode.procedure.StateMachineProcedure;
+import org.apache.iotdb.confignode.procedure.env.ConfigNodeProcedureEnv;
+import org.apache.iotdb.confignode.procedure.exception.ProcedureException;
+import org.apache.iotdb.confignode.procedure.scheduler.SimpleProcedureScheduler;
+import org.apache.iotdb.confignode.procedure.state.ProcedureLockState;
+import org.apache.iotdb.confignode.procedure.state.RemoveConfigNodeState;
+import org.apache.iotdb.confignode.procedure.store.ProcedureFactory;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.DataOutputStream;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+
+/** remove config node procedure */
+public class RemoveConfigNodeProcedure
+    extends StateMachineProcedure<ConfigNodeProcedureEnv, RemoveConfigNodeState> {
+  private static final Logger LOG = LoggerFactory.getLogger(RemoveConfigNodeProcedure.class);
+  private static final int retryThreshold = 5;
+
+  private TConfigNodeLocation tConfigNodeLocation;
+
+  public RemoveConfigNodeProcedure() {
+    super();
+  }
+
+  public RemoveConfigNodeProcedure(TConfigNodeLocation tConfigNodeLocation) {
+    super();
+    this.tConfigNodeLocation = tConfigNodeLocation;
+  }
+
+  @Override
+  protected Flow executeFromState(ConfigNodeProcedureEnv env, RemoveConfigNodeState state) {
+    if (tConfigNodeLocation == null) {
+      return Flow.NO_MORE_STATE;
+    }
+    try {
+      switch (state) {
+        case REMOVE_CONFIG_NODE_PREPARE:
+          setNextState(RemoveConfigNodeState.REMOVE_PEER);
+          break;
+        case REMOVE_PEER:
+          env.removeConfigNodePeer(tConfigNodeLocation);
+          setNextState(RemoveConfigNodeState.REMOVE_CONSENSUS_GROUP);
+          LOG.info("Remove peer {}", tConfigNodeLocation);
+          break;
+        case REMOVE_CONSENSUS_GROUP:
+          env.removeConsensusGroup(tConfigNodeLocation);
+          setNextState(RemoveConfigNodeState.STOP_CONFIG_NODE);
+          LOG.info("Remove Consensus Group {}", tConfigNodeLocation);
+          break;
+        case STOP_CONFIG_NODE:
+          env.stopConfigNode(tConfigNodeLocation);
+          LOG.info("Stop Config Node {}", tConfigNodeLocation);
+          return Flow.NO_MORE_STATE;
+      }
+    } catch (Exception e) {
+      if (isRollbackSupported(state)) {
+        setFailure(new ProcedureException("Remove Config Node failed " + state));
+      } else {
+        LOG.error(
+            "Retrievable error trying to remove config node {}, state {}",
+            tConfigNodeLocation,
+            state,
+            e);
+        if (getCycles() > retryThreshold) {
+          setFailure(new ProcedureException("State stuck at " + state));
+        }
+      }
+    }
+    return Flow.HAS_MORE_STATE;
+  }
+
+  @Override
+  protected void rollbackState(ConfigNodeProcedureEnv env, RemoveConfigNodeState state)
+      throws IOException, InterruptedException, ProcedureException {}
```

Comment:
Is the `rollbackState` method of class `RemoveConfigNodeProcedure` needed to be implemented?

Reply:
Thanks.
Because there was such a scene. For example, if a node breaks down and cannot be recovered, you need to delete the node. In this case, you can only perform the first step of removing peer on the leader side. The nodes to be removed cannot communicate with each other. In this case, the operations in the second and third steps will fail. In this case, the rollback cannot be performed.